### ISSUE TITLE
The counter example from tutorial doesn't work with `St` annotations

### ISSUE THREAD

**Issue by cpitclaudel** [FStarLang/FStar#1056]
The tutorial says this:

> Sometimes it is useful to return a function after performing some computation. For example, here is a function that makes a new counter, initializing it to init.
> 
>     let new_counter init =
>       let c = ST.alloc init in
>       fun () -> c := !c + 1; !c
> 
> F* infers that new_counter has type int -> ST (unit -> ST int), the type of a function that given an integer, may read, write or allocate references, and then returns a unit -> ST unit function. You can write down this type and have F* check it; or, using sub-effecting, you can write the type below and F* will check that new_counter has that type as well. 

But when I run the function through F*, I get this:

> Bound variables 'c#822610' escapes; add a type annotation

(As a side note, there's probably a typo in escape*s* or variable*s*)

**Comment by cpitclaudel** [comment]
Running this in the online editor next to the tutorial gives me this instead:

```
/var/fstar/ulib/FStar.ST.fst(26,13-26,21) : (Error) Identifier not found: [st_pre_h] 
```

**Comment by catalin-hritcu** [comment]
@cpitcludel  I already filed online tutorial problem yesterday (#1052).

As for the rest, it seems like our tutorial is getting more and more obsolete (provided this was ever correct). Please help make this better.

**Comment by catalin-hritcu** [comment]
I've changed the tutorial to add a type annotation, since as Clement's first error message says we need to at least `add a type annotation`. As @kyoDralliam noticed today though, there are bigger problems with this example. It doesn't even work with the type annotation:
```
val new_counter : int -> St (unit -> St int)
let new_counter init =
  let c = alloc init in
  fun () -> c := !c + 1; !c
```
The error message:
```
(Error) assertion failed(Also see: /home/hritcu/Projects/fstar/pub/ulib/FStar.ST.fst(31,26-31,40))
```
points at `lift_div_state` in `FStar.ST.fst`:
```
unfold let lift_div_state (a:Type) (wp:pure_wp a) (p:st_post a) (h:heap) = wp (fun a -> p a h)
```
Funny enough, if we replace `St` by `ML` in the type this does work:
```
val new_counter_ml : int -> ML (unit -> ML int)
```
This seems like a bug to me so marking as such.

**Comment by nikswamy** [comment]
This example now needs an annotation to infer the preorder.

```
module Test
open FStar.Heap
open FStar.ST

val new_counter : int -> St (unit -> St int)
let new_counter init =
  let c : ref int = alloc init in
  fun () -> c := !c + 1; !c
```

It still fails with 
```
.\test.fst(6,0-8,27): (Error) could not prove post-condition
```

@aseemr : I was expecting this to be resolved by the support you added in the SMT encoding for representing impure function types (which was previously a very lossy part of the encoding).

**Comment by aseemr** [comment]
1. There is an `FStar.Ref` module now that one can use to work with references with default preorder. This should eliminate the need for the annotation.

2. I haven't still fixed the encoding of impure arrows. Will keep it in my list for this week. (It is also tracked via #855).

**Comment by jmitchell** [comment]
I also stumbled on this using v0.9.6.0.

Are there other small `St` examples that work?

**Comment by catalin-hritcu** [comment]
You can have a look at these slides: http://prosecco.gforge.inria.fr/personal/hritcu/teaching/epit2018/slides/out/state/state.html#/sec-verifying-stateful-programs-with-f-

They are from this summer school: http://prosecco.gforge.inria.fr/personal/hritcu/teaching/epit2018/

**Comment by aseemr** [comment]
The `St` as well as `ML` examples work after the fix to #855 :

```
open FStar.ST
open FStar.All

val new_counter : int -> ML (unit -> ML int)
let new_counter init =
    let c = alloc init in
    fun () -> c := !c + 1; !c
```

and

```
module Test

open FStar.Heap
open FStar.Ref

val new_counter : ℤ → St (unit → St ℤ)
let new_counter init =
  let c : ref ℤ = alloc init in
  λ () → c ≔ !c + 1; !c
```

Closing, thanks for the repros and the discussions.
